### PR TITLE
[Formula] Performance investigation attempt #1

### PR DESCRIPTION
**Pull request by dej611**
## Summary

Tried to profile the Formula component and found out some possible bottlenecks.
Here's the full list of bottlenecks from my research:
* `TopNavBar` re-render on formula typing => addressed partially here 📝 
* `AutosizeInput` DOM restyle => addressed partially here 📝 
* `ConfigPanel` re-render => addressed here 📝 
* `SuggestionPanel` computation + render => reduced some looping here, nothing super important 📝 
* Monaco `suggestion` box style computation
* Monaco initialization => removed some unused operation 📝 

The profiling process has been conducted with a 4x and 6x CPU slowdown.

The `AutoresizeInput` is relative to the EUI `ComboBox` component, which gets measured over and over on re-render. The `updateInputWidth` method in it requires at least 50ms of execution every time.

The Monaco editor `suggestion` box has to compute the right layout every time is triggered, leading to very consuming tasks when typing - as it has to adapt the position based on the cursor. I had a look at the Monaco API but couldn't find anything particuarly useful for that.

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- [ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)
- [ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

### CODE REVIEW

File: x-pack/plugins/lens/public/app_plugin/lens_top_nav.tsx [hunk 1]
```diff
@@ -164,124 +164,163 @@ export const LensTopNavMenu = ({
...
-                content: exporters.datatableToCSV(datatable, {
-                  csvSeparator: uiSettings.get('csv:separator', ','),
-                  quoteValues: uiSettings.get('csv:quoteValues', true),
-                  formatFactory: data.fieldFormats.deserialize,
-                }),
-                type: exporters.CSV_MIME_TYPE,
-              };
+                  memo[`${lastKnownDoc?.title || unsavedTitle}${postFix}.csv`] = {
+                    content: exporters.datatableToCSV(datatable, {
+                      csvSeparator: uiSettings.get('csv:separator, ,'),
```
Comment (flash1293) [inline x-pack/plugins/lens/public/app_plugin/lens_top_nav.tsx]: This change doesn't look right, did it happen by accident?

Comment (dej611) [reply x-pack/plugins/lens/public/app_plugin/lens_top_nav.tsx]: I think prettier just went wild here. I'll restore it

**Comment by dej611** [comment]
On any state change the `AutosizeInput` is generating huge delays on my machine:

Screenshot 2021-06-07 at 17 03 52
Screenshot 2021-06-07 at 17 04 30

**Comment by flash1293** [comment]
@dej611 Should we raise this on the EUI side? Seems like a general problem

**Comment by flash1293** [comment]
Not sure how it would be so slow - this is the code: https://github.com/JedWatson/react-input-autosize/blob/8a68b453ca1dec723e255c40eab7f2928228fe22/src/AutosizeInput.js#L103

**Comment by dej611** [comment]
This is the offending line: `typeof this.sizer.scrollWidth === 'undefined'`

That line it causing a reflow (ref: https://gist.github.com/paulirish/5d52fb081b3570c81e3a ) and it's called at every single state change.

**Comment by dej611** [comment]
With no slowdown the resources required by the `scrollWidth` reading is mostly negligible (still it's the most expensive task in the profiling):


Screenshot 2021-06-07 at 17 55 36


Same test by with a 4x slowdown and it gets much worst:

Screenshot 2021-06-07 at 17 52 38


The only component in view which is using it is the `FormatSelector`. At this point I think it's worth raise the problem upstream and live with it for now.

The test scenario I used is the following:
* Open the dimension panel
* Click on Formula
* Manually type the formula: `count(kql='category.keyword : "Men\'s Clothing" ') + average(products.price)` (Used autocomplete only for the value)
* Close the panel
* Stop the profiling

**Comment by dej611** [comment]
Managed to reduced a bit.

**Comment by dej611** [comment]
With latest changes writing a formula, even with the 4x slowdown, feels less sluggish.
The only exception is typing a KQL/Lucene query: in that case the `showSuggestions` in Monaco kicks in makes everything slow again. As soon as the cursor gets out of the query area, it's fast again.

**Pull request merged**